plottr/gui/widgets.py

```diff
@@ -305,3 +306,183 @@ def _onButton(self) -> None:
         else:
             self.widget.setVisible(False)
             self.btn.setText(self.collapsedTitle)
+
+
+class DimensionCombo(QtWidgets.QComboBox):
+    """A Combo Box that allows selection of a single data dimension.
+    This widget is designed to be used in a node widget.
+
+    Which type of dimensions are available for selection is set through the
+    ``dimensionType`` option when creating the instance.
+
+    The widget can be linked to a node using the :meth:`.connectNode` method.
+    After linking, the available options will be populated whenever the data in
+    the node changes.
+    """
+
+    #: Signal(str)
+    #: emitted when the user selects a dimension.
+    dimensionSelected = Signal(str)
+
+    def __init__(self, parent: Optional[QtWidgets.QWidget] = None,
+                 dimensionType: str = 'axes') -> None:
+        """Constructor.
+
+        :param parent: parent widget
+        :param dimensionType: one of `axes`, `dependents` or `all`.
+        """
+        super().__init__(parent)
+
+        self.node: Optional[Node] = None
+        self.dimensionType = dimensionType
+
+        self.clear()
+        self.entries = ['None']
+        for e in self.entries:
+            self.addItem(e)
+
+        self.currentTextChanged.connect(self.signalDimensionSelection)
+
+    def connectNode(self, node: Optional[Node] = None) -> None:
+        """Connect a node. will result in populating the combo box options
+        based on dimensions available in the node data.
+
+        :param node: instance of :class:`.Node`
+        """
+        if node is None:
+            raise RuntimeError
+        self.node = node
+        if self.dimensionType == 'axes':
+            self.node.dataAxesChanged.connect(self.setDimensions)
+        elif self.dimensionType == 'dependents':
+            self.node.dataDependentsChanged.connect(self.setDimensions)
+        else:
+            self.node.dataFieldsChanged.connect(self.setDimensions)
+
+    @updateGuiQuietly
+    def setDimensions(self, dims: Sequence[str]) -> None:
+        """Set the dimensions that are available for selection.
+
+        :param dims: list of dimensions, as strings.
+        :return: ``None``
+        """
+        self.clear()
+        allDims = self.entries + list(dims)
```

Comment:
should this actually change self.entries and call addItem on elements of that? or is self.entries actually unused?

Reply:
``self.entries`` contains currently only one element, ``None``, to give the user the option to not select any item.
And that entry should be stored somewhere other than the actual item list, because it would be cleared when we set a new list of dimensions.
So there will in any case always at least be the ``None`` in the dropdown. It looks maybe a bit weird, but i wanted to have that information not in the function but rather in the constructor, and was thinking of cases where more than just the ``None`` could be useful to fall under that category.